devito/data/data.py

```diff
@@ -260,6 +388,176 @@ def _convert_index(self, glb_idx):
 
         return loc_idx[0] if len(loc_idx) == 1 else tuple(loc_idx)
 
+    def _set_global_idx(self, val, idx, val_idx):
+        data_loc_idx = as_tuple(val._convert_index(val_idx))
+        data_global_idx = []
+        # Convert integers to slices so that shape dims are preserved
+        if is_integer(as_tuple(idx)[0]):
+            data_global_idx.append(slice(0, 1, 1))
+        for i, j in zip(data_loc_idx, val._decomposition):
+            if not j.loc_empty:
+                data_global_idx.append(j.convert_index_global(i))
+            else:
+                data_global_idx.append(None)
+        mapped_idx = []
+        for i, j in zip(data_global_idx, as_tuple(idx)):
+            if isinstance(j, slice) and j.start is None:
+                norm = 0
+            elif isinstance(j, slice) and j.start is not None:
+                norm = j.start
+            else:
+                norm = j
+            if i is not None:
+                if isinstance(j, slice) and j.step is not None:
+                    mapped_idx.append(slice(j.step*i.start+norm,
+                                            j.step*i.stop+norm, j.step))
+                else:
+                    mapped_idx.append(slice(i.start+norm, i.stop+norm, i.step))
+            else:
+                mapped_idx.append(None)
+        return as_tuple(mapped_idx)
+
+    def _local_data_idx(self, loc_idx):
+        loc_data_idx = []
+        for i in as_tuple(loc_idx):
+            if isinstance(i, slice) and i.step is not None and i.step == -1:
+                if i.stop is None:
+                    loc_data_idx.append(slice(0, i.start+1, -i.step))
+                else:
+                    loc_data_idx.append(slice(i.stop+1, i.start+1, -i.step))
+            elif isinstance(i, slice) and i.step is not None and i.step < -1:
+                if i.stop is None:
+                    lmin = i.start
+                    while lmin >= 0:
+                        lmin += i.step
+                    loc_data_idx.append(slice(lmin-i.step, i.start+1, -i.step))
+                else:
+                    loc_data_idx.append(slice(i.stop+1, i.start+1, -i.step))
+            elif is_integer(i):
+                loc_data_idx.append(slice(i, i+1, 1))
+            else:
+                loc_data_idx.append(i)
+        return as_tuple(loc_data_idx)
+
+    def _mpi_index_maps(self, loc_idx, val_shape, comm, rank):
```

Comment:
this could be a free function 

self isn't actually necessary:
* nprocs -> can be obtained from ``comm``
* topology and all_cords can be passed explicitly